module/zfs/zfs_ioctl.c

```diff
@@ -3267,6 +3266,18 @@ zfs_ioc_create(const char *fsname, nvlist_t *innvl, nvlist_t *outnvl)
 	if (cbfunc == NULL)
 		return (SET_ERROR(EINVAL));
 
+	/* can't create below anything but filesystems (eg. no ZVOLs) */
+	zfs_get_parent(fsname, parentname, sizeof (parentname));
+	error = dmu_objset_hold(parentname, FTAG, &parentos);
+	if (error != 0) {
+		return (error);
+	}
+	if (dmu_objset_type(parentos) != DMU_OST_ZFS) {
+		dmu_objset_rele(parentos, FTAG);
+		return (SET_ERROR(ENOTDIR));
```

Comment:
I think it would be better to put this check in `dmu_objset_create_check()`, for 2 reasons:
1. It looks like this check is racy - after calling dmu_objset_rele(), the the parent could be renamed and replaced with a zvol.
2. Although there are no other callers of dmu_objset_create_check() today, it's logically the right place to make any checks that should always be enforced.  For example, we would like to add objset-creation to channel programs, and that would use dmu_objset_create_check/sync directly.

Reply:
That stack doesn't make any sense to me, but the panic message does - you can't call dsl_pool_config_enter() from a sync task (checkfunc or syncfunc).  But you shouldn't need to.

Reply:
I just moved this part of the code from "zfs_ioctl.c" (`zfs_ioc_create()`) to "dmu_objset.c" (`dmu_objset_create_check()`).

Relevant part of the diff:

```diff
diff --git a/module/zfs/dmu_objset.c b/module/zfs/dmu_objset.c
index a8304c18d..99308d9d6 100644
--- a/module/zfs/dmu_objset.c
+++ b/module/zfs/dmu_objset.c
@@ -1118,6 +1119,8 @@ dmu_objset_create_check(void *arg, dmu_tx_t *tx)
        dmu_objset_create_arg_t *doca = arg;
        dsl_pool_t *dp = dmu_tx_pool(tx);
        dsl_dir_t *pdd;
+       objset_t *parentos;
+       char parentname[ZFS_MAX_DATASET_NAME_LEN];
        const char *tail;
        int error;
 
@@ -1147,6 +1150,21 @@ dmu_objset_create_check(void *arg, dmu_tx_t *tx)
        error = dsl_fs_ss_limit_check(pdd, 1, ZFS_PROP_FILESYSTEM_LIMIT, NULL,
            doca->doca_cred);
 
+       /* can't create below anything but filesystems (eg. no ZVOLs) */
+       zfs_get_parent(doca->doca_name, parentname, sizeof (parentname));
+       error = dmu_objset_hold(parentname, FTAG, &parentos);
+       if (error != 0) {
+               dsl_dir_rele(pdd, FTAG);
+               return (error);
+       }
+
+       if (dmu_objset_type(parentos) != DMU_OST_ZFS) {
+               dmu_objset_rele(parentos, FTAG);
+               dsl_dir_rele(pdd, FTAG);
+               return (SET_ERROR(ENOTDIR));
+       }
+       dmu_objset_rele(parentos, FTAG);
+
        dsl_dir_rele(pdd, FTAG);
 
        return (error);
```

Maybe i can't `dmu_objset_hold()` the `parentos` in syncing context? Should i just put this code inside a `if (!dmu_tx_is_syncing(tx))` code block?

Reply:
Correct, you can not use dmu_objset_hold().  You should use dsl_dataset_hold() or dsl_dir_hold() instead.  dmu_objset_create_check() is already doing this for you, so you should be able to use `pdd` (the parent dsl_dir) to check the parent's type.

Reply:
As you have seen, you can get the type from dmu_objset_type().  So you just need to get the objset given the dsl_dir.  I think you'll want something like `dmu_objset_from_ds(dsl_dataset_open_obj(pdd->phys->dd_head_dataset_obj))` (not literally)